### ISSUE TITLE
Introduce `tokio::start` and deprecate `tokio::run`

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio#840]
Currently, `tokio::run` combines two concerns:

* Starting the runtime.
* Spawning a future.

The function takes a future as an argument and blocks until the future and all tasks spawned as a result of that future complete. However, the name can be confusing in that `tokio::run(future)` can be interpreted as running the specific future.

There also is another problem where the future passed to `run` is constructed without having any knowledge of the runtime. This causes problems when the construction process needs to be able to access runtime details. This problem will get worse with tokio-trace.

To work around this issue, currently it is common to do the following:

```rust
tokio::run(future::lazy(|| {
    tokio::spawn(construct_future());
    Ok(())
}));
```

# Proposal

A new function is introduced:

```rust
fn start<F: FnOnce() -> R, R>(init: F) -> R { ... } 
```

This function would start the runtime and run the provided closure in context of the runtime.

In order to replicate `tokio::run`, the following can be done:

```rust
tokio::start(|| {
    tokio::spawn(construct_future());
});
```

# Drawbacks

The proposal does not include a way to block the current thread until a future is complete and get the value without constructing an explicit runtime instance. So, the following would be necessary:

```rust
let mut rt = Runtime::new().unwrap();
let res = rt.block_on(my_future);
```

**Comment by davidbarsky** [comment]
I'm in favor of this change. With the introduction of this change, would it be helpful to introduce corresponding `tokio::start_blocking` for testing purposes and thereby addressing the drawbacks?

**Comment by LucioFranco** [comment]
### Start / Run

I am in favor of this change but I personally would like to suggest that we allow the bootstrap closure to return a future that will get spawned. This follows how futures lazy works and would align with how people already use `tokio::run`. It seems that seastar uses the same concept [here](http://docs.seastar.io/master/classseastar_1_1app__template.html) in that the closure passed to the run function can return a future.

I looked through some of my own tokio codebases and I find generally that servers and clients will like to return one future as the main one even if spawning one and returning `Ok(())` works.

- [Here](https://github.com/LucioFranco/tower-consul/blob/master/examples/hyper.rs#L9) is an example within my consul library
- Another example within toshi-search/Toshi [here](https://github.com/toshi-search/Toshi/blob/984475e92646fd1ac4344e6029ec45e2f438c22d/src/bin/toshi-placement.rs#L30)
- The last example I have from my own code is also [here](https://github.com/LucioFranco/clique/blob/master/clique-agent/src/main.rs#L27)

So what I suggest is that we match how `future::lazy` does it and have the tokio run function look like so:

```rust
fn run<F, R>(init: F) -> R
where
    F: FnOnce() -> R,
    R: IntoFuture<Item = (), Error = ()>,
{ ... }
```
(I am omiting some trait bounds like `Send` and `'static` but those are trivial)

This would in turn change the examples above like so:

```rust
tokio::run(|| construct_future());
```

Or

```rust
tokio::run(|| {
	tokio::spawn(some_bg_task());
	tokio::spawn(some_other_bg_task());
	
	serve()
});
```

Personally, this aligns with how I think about servers and code in general as there is some single entry point like `serve` in this example.

I also think we should update `run` and to not using `start`, as this would be a positive breaking change and one that would be easy to update to since tokio code bases don't generally have a large amount of `tokio::run` calls. If we are ever going to update `run` it is going to have to be sooner rather than later. We can also move the current run to a deprecated version to also ease the update path.

### Block On

As for block_on per the discussion that @carllerche and I had today on gitter. It seems that there should be a new crate `tokio-test` that will contain a `tokio_test::block_on`. I think it is a good idea to leave out a convenience function for `block_on` as its generally not a good idea to create a lot of tokio runtimes when the program could probably be refactored to live as a tree of futures.

**Comment by LucioFranco** [comment]
@carllerche I spent more time thinking about this today and I actually am now very much in favor of not returning an `IntoFuture`. I think this will cause less indirection. I also am backtracking on my comment about renaming it from `start` to `run` because `tokio::run(serve)` would solve what I wanted and `tokio::start` could be used for more advanced applications because my original issue was that `tokio::run(|| tokio::spawn(serve))` is not very ergonomic.

So TL;DR I am very much in favor of your original proposal and I think this will be a really good positive change for newer tokio users. 👍

**Comment by tobz** [comment]
Finally reading this over: am I understanding correctly that while the closure given to `tokio::start` would not need to return a `Future`/`IntoFuture`, `tokio::start` would still block on the runtime going idle?

**Comment by stjepang** [comment]
> `tokio::start` would still block on the runtime going idle?

Correct. `tokio::start(...)` is equivalent to:

```rust
tokio::run(future::lazy(|| {
    tokio::spawn(...);
    Ok(())
}))
```

**Comment by carllerche** [comment]
`tokio::run` will be removed in favor of a proc macro:

```rust
#[tokio::main]
async fn main() {
}
```
